java/src/jni/h5util.c

```diff
@@ -4079,6 +4081,521 @@ translate_rbuf(JNIEnv *env, jobjectArray ret_buf, jlong mem_type_id, H5T_class_t
+            void *objBuf = NULL;
+
+            /* Convert each compound element */
+            int nmembs = H5Tget_nmembers(mem_type_id);
+
+            /* invoke the toArray method */
+            if (mToArray == NULL)
+                CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+            jobjectArray array   = (jobjectArray)ENVPTR->CallObjectMethod(ENVONLY, in_obj, mToArray);
+            jsize        jnelmts = ENVPTR->GetArrayLength(ENVONLY, array);
+
+            if (jnelmts != nmembs)
+                H5_BAD_ARGUMENT_ERROR(
+                    ENVONLY, "translate_atomic_wbuf: number of elements not equal to number of members");
+
+            /* Convert each compound object to an element */
+            for (i = 0; i < (size_t)nmembs; i++) {
+                H5T_class_t memb_vlClass;
+                size_t      memb_vlSize;
+                size_t      memb_offset;
+
+                if ((memb = H5Tget_member_type(mem_type_id, (unsigned int)i)) < 0)
+                    H5_LIBRARY_ERROR(ENVONLY);
+                if ((memb_offset = H5Tget_member_offset(mem_type_id, (unsigned int)i)) < 0)
+                    H5_LIBRARY_ERROR(ENVONLY);
+                if ((memb_vlClass = H5Tget_class(memb)) < 0)
+                    H5_LIBRARY_ERROR(ENVONLY);
+                if (!(memb_vlSize = H5Tget_size(memb)))
+                    H5_LIBRARY_ERROR(ENVONLY);
+
+                jobject arr_obj = ENVPTR->GetObjectArrayElement(ENVONLY, array, i);
+                translate_atomic_wbuf(ENVONLY, arr_obj, memb, memb_vlClass,
+                                      ((char *)raw_buf) + i * typeSize + memb_offset);
+                ENVPTR->DeleteLocalRef(ENVONLY, arr_obj);
+                H5Tclose(memb);
+            }
+            break;
+        } /* H5T_COMPOUND */
+        case H5T_ARRAY: {
+            void *objBuf = NULL;
+
+            if (!(memb = H5Tget_super(mem_type_id)))
+                H5_LIBRARY_ERROR(ENVONLY);
+            if ((vlClass = H5Tget_class(memb)) < 0)
+                H5_LIBRARY_ERROR(ENVONLY);
+            if (!(vlSize = H5Tget_size(memb)))
+                H5_LIBRARY_ERROR(ENVONLY);
+
+            /* Convert each array element */
+            /* invoke the toArray method */
+            if (mToArray == NULL)
+                CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+            jobjectArray array   = (jobjectArray)ENVPTR->CallObjectMethod(ENVONLY, in_obj, mToArray);
+            jsize        jnelmts = ENVPTR->GetArrayLength(ENVONLY, array);
+
+            if (jnelmts < 0)
+                H5_BAD_ARGUMENT_ERROR(ENVONLY, "translate_atomic_wbuf: number of array elements < 0");
+
+            if (NULL == (objBuf = HDmalloc((size_t)jnelmts * vlSize)))
+                H5_OUT_OF_MEMORY_ERROR(ENVONLY, "translate_atomic_wbuf: failed to allocate buffer");
+
+            translate_wbuf(ENVONLY, array, memb, vlClass, (jsize)jnelmts, objBuf);
+
+            HDmemcpy((char *)raw_buf + vlSize * jnelmts, (char *)objBuf, vlSize * jnelmts);
```

Comment:
Same here with the addition of `vlSize * jnelmts`?

Reply:
Might be a holdover from when the loops were there before being moved.